src/stats/pca.jl

```diff
@@ -0,0 +1,160 @@
+# Implements Candid Covariance-Free Incremental PCA algorithm of Weng et al:
+# Weng, Juyang, Yilu Zhang, and Wey-Shiuan Hwang. "Candid covariance-free 
+# incremental principal component analysis." IEEE Transactions on Pattern 
+# Analysis and Machine Intelligence 25.8 (2003): 1034-1040.
+#
+# A comparison of online PCA algorithms in 2018 showed it to be a good
+# trade-off in that it is fast and has low reconstruction error:
+# H. Cardot and D. Degras, "Online Principal Component Analysis in 
+# High Dimension: Which Algorithm to Choose?", Int. Statistical Review, 2018-
+
+"""
+    CCIPCA(indim::Int, outdim::Int; l::Int)
+
+Online PCA with the CCIPCA (Candid Covariance-free Incremental PCA) algorithm,
+where indim is the length of incoming vectors, outdim is the number of 
+dimension to project to, and l is the level of amnesia. Give values of l in 
+the range 2-4 if you want old vectors to be gradually less important, i.e. 
+latest vectors added get more weight.
+
+This is a very fast, simple, and online approximation of PCA. It can be used
+for Dimensionality Reduction to project high-dimensional vectors into a 
+low-dimensional (typically 2D or 3D) space. This algorithm has shown very 
+good properties in comparative studies; it is both fast and give a good 
+approximation to (batch) PCA.
+
+# Example
+
+    o = CCIPCA(10, 2)                # Project 10-dimensional vectors into 2D
+    u1 = rand(10)
+    fit!(o, u1)                      # Fit to u1
+    u2 = rand(10)
+    fit!(o, u2)                      # Fit to u2
+    u3 = rand(10)
+    OnlineStats.transform(o, u3)     # Project u3 into PCA space fitted to u1 and u2 but don't change the projection
+    u4 = rand(10)
+    OnlineStats.fittransform!(o, u4) # Fit u4 and then project u4 into the space
+    sort!(o)                         # Sort from high to low eigenvalues
+    o[1]                             # Get primary (1st) eigenvector
+    OnlineStats.variation(o)         # Get the variation (explained) "by" each eigenvector
+"""
+mutable struct CCIPCA <: OnlineStat{Vector{Float64}}
+    U::Matrix{Float64}      # Eigenvectors, one row per indim, one column per outdim
+    lambda::Vector{Float64} # Eigenvalues, one per outdim
+    center::Vector{Float64} # Center/mean, one per indim
+    l::Int
+    n::Int
+end
+function CCIPCA(indim::Int, outdim::Int; l::Int=0)
```

Comment:
I'll leave it for you to decide, but for things like `CovMatrix` I determine the `indim` from the first time `_fit!` is called. 

```julia
fit!(CovMatrix(), eachcol(x))
```

is a nice convenience vs.

```julia
fit!(CovMatrix(size(x, 1)), eachcol(x))
```

Reply:
Hmm, I guess I could set it to a default size and then update it on first call to _fit if size differs.